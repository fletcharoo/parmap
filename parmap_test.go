package parmap_test

import (
	"fmt"
	"os"
	"testing"

	"github.com/fletcharoo/parmap"
	"github.com/gkampitakis/go-snaps/snaps"
)

func TestMain(m *testing.M) {
	r := m.Run()
	snaps.Clean(m, snaps.CleanOpts{Sort: true})
	os.Exit(r)
}

var (
	_errMap = parmap.ErrMap{
		1: fmt.Errorf("error 1"),
		2: fmt.Errorf("error 2"),
		3: fmt.Errorf("error 3"),
	}
	_inputs = []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20}
)

func Test_ErrMap_ErrJoin(t *testing.T) {
	snaps.MatchSnapshot(t, _errMap.ErrJoin())
}

func Test_ErrMap_String(t *testing.T) {
	snaps.MatchSnapshot(t, fmt.Sprint(_errMap.ErrJoin()))
}

func Test_Do(t *testing.T) {
	do := func(i int) (int, error) {
		return i * 5, nil
	}

	result, err := parmap.Do(_inputs, do)

	if err != nil {
		t.Fatalf("Unexpected error: %s", err)
	}

	snaps.MatchJSON(t, result)
}

func Test_Do_Error(t *testing.T) {
	do := func(i int) (int, error) {
		if i%3 == 0 {
			return 0, fmt.Errorf("error %d", i)
		}

		return i, nil
	}

	result, err := parmap.Do(_inputs, do)

	t.Run("Error", func(t *testing.T) {
		snaps.MatchSnapshot(t, err)
	})

	t.Run("Result", func(t *testing.T) {
		snaps.MatchSnapshot(t, result)
	})
}
