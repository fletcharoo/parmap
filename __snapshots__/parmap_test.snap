
[Test_Do - 1]
[
 5,
 10,
 15,
 20,
 25,
 30,
 35,
 40,
 45,
 50,
 55,
 60,
 65,
 70,
 75,
 80,
 85,
 90,
 95,
 100
]
---

[Test_Do_Error/Error - 1]
parmap.ErrMap{
    2:  &errors.errorString{s:"error 3"},
    5:  &errors.errorString{s:"error 6"},
    8:  &errors.errorString{s:"error 9"},
    11: &errors.errorString{s:"error 12"},
    14: &errors.errorString{s:"error 15"},
    17: &errors.errorString{s:"error 18"},
}
---

[Test_Do_Error/Result - 1]
[]int{1, 2, 0, 4, 5, 0, 7, 8, 0, 10, 11, 0, 13, 14, 0, 16, 17, 0, 19, 20}
---

[Test_ErrMap_ErrJoin - 1]
&errors.joinError{
    errs: {
        &fmt.wrapError{
            msg: "1: error 1",
            err: &errors.errorString{s:"error 1"},
        },
        &fmt.wrapError{
            msg: "2: error 2",
            err: &errors.errorString{s:"error 2"},
        },
        &fmt.wrapError{
            msg: "3: error 3",
            err: &errors.errorString{s:"error 3"},
        },
    },
}
---

[Test_ErrMap_String - 1]
1: error 1
2: error 2
3: error 3
---
